flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultExecutionSlotAllocator.java

```diff
@@ -0,0 +1,294 @@
+			return CompletableFuture.completedFuture(preferredLocationsBasedOnState);
+		}
+
+		return getPreferredLocationsBasedOnInputs(executionVertexId, inputsLocationsRetriever);
+	}
+
+	/**
+	 * Gets the location preferences of the execution, as determined by the locations
+	 * of the predecessors from which it receives input data.
+	 * If there are more than {@link MAX_DISTINCT_LOCATIONS_TO_CONSIDER} different locations of source data,
+	 * or neither the sources have not been started nor will be started with the execution together,
+	 * this method returns an empty collection to indicate no location preference.
+	 *
+	 * @return The preferred locations based in input streams, or an empty iterable,
+	 *         if there is no input-based preference.
+	 */
+	@VisibleForTesting
+	static CompletableFuture<Collection<TaskManagerLocation>> getPreferredLocationsBasedOnInputs(
+			ExecutionVertexID executionVertexId,
+			InputsLocationsRetriever inputsLocationsRetriever) {
+		CompletableFuture<Collection<TaskManagerLocation>> preferredLocations = null;
+
+		Collection<CompletableFuture<TaskManagerLocation>> locationsFutures = new ArrayList<>();
+
+		Collection<Collection<ExecutionVertexID>> allProducers =
+				inputsLocationsRetriever.getConsumedResultPartitionsProducers(executionVertexId);
+		for (Collection<ExecutionVertexID> producers : allProducers) {
+			Set<TaskManagerLocation> inputLocations = new HashSet<>();
+			List<CompletableFuture<TaskManagerLocation>> inputLocationsFutures = new ArrayList<>();
+
+			for (ExecutionVertexID producer : producers) {
+				Optional<CompletableFuture<TaskManagerLocation>> optionalLocationFuture =
+						inputsLocationsRetriever.getTaskManagerLocation(producer);
+				optionalLocationFuture.ifPresent(
+						(locationFuture) -> {
+							locationsFutures.add(locationFuture);
+							TaskManagerLocation taskManagerLocation = locationFuture.getNow(null);
+							if (taskManagerLocation != null) {
+								inputLocations.add(taskManagerLocation);
+							} else {
+								inputLocationsFutures.add(locationFuture);
+							}
+						}
+				);
+				// If the parallelism is large, wait for all futures coming back may cost a long time,
+				// so using the unfulfilled future number here to speed up it.
+				if (inputLocations.size() > MAX_DISTINCT_LOCATIONS_TO_CONSIDER ||
+						inputLocationsFutures.size() > MAX_DISTINCT_LOCATIONS_TO_CONSIDER) {
+					locationsFutures.clear();
+					break;
+				}
+			}
+
+			if (!locationsFutures.isEmpty()) {
+				CompletableFuture<Collection<TaskManagerLocation>> uniqueLocationsFuture =
+						FutureUtils.combineAll(locationsFutures).thenApply(
+								(locations) -> {
+									Set<TaskManagerLocation> uniqueLocations = new HashSet<>(locations);
+									if (uniqueLocations.size() <= MAX_DISTINCT_LOCATIONS_TO_CONSIDER) {
+										return uniqueLocations;
+									} else {
+										return Collections.emptySet();
+									}
+								});
+				if (preferredLocations == null) {
```

Comment:
null check not needed if you initialize `preferredLocations`

Reply:
Done